### DISCUSSION TITLE
Subgraph dataset implementation

### DISCUSSION THREAD

**Discussion by dongkwan-kim** [pyg-team/pytorch_geometric#2728] [Ideas]
Hi, I am implementing codes for subgraph datasets in [Subgraph Neural Networks](https://proceedings.neurips.cc/paper/2020/hash/5bca8566db79f3788be9efd96c9ed70d-Abstract.html). I need some advices before creating a PR.

In `InMemoryDataset`, we first create `data_list` and call `collate` to get `Data` and `slices`. However, in subgraph prediction tasks, there are one supergraph (edges `edge_index` and node features `x`) and a set of subgraphs (node indices `x_index`). Edges in subgraphs can be induced from the supergraph using `utils.subgraph`.

Since edges in the global graph are not given to each subgraph instance, we cannot collate them into a single `Data` object. In my latest version, I made two separate member variables in `InMemoryDataset`: `self.global_data` for the supergraph, `self.data` for  ​a set of subgraphs (w/ `self.collate`). 

There does not seem to be an example of subgraph datasets in PyG (I might miss them), so I would like to discuss what the best form will be. Thanks in advance.

**Comment by rusty1s** [comment]
This depends on where you actually want to do the subgraph processing. I think you have two options:
1. Store subgraphs in separate `data` objects and collate them afterwards (good runtime, but bad memory consumption)
2. Store a single data object via `self.collate`, and override `len` and `get` in `InMemoryDataset`:
```python
@property
def len(self):
    return x  # The number of subgraphs

def get(self, idx):
    # Perform subgraph computation here and return the sub data.
```

**Reply by dongkwan-kim** [reply]
I see. Thank you for the suggestion.
It would be nice if we let the users choose the processing method for their own use cases.
I will send the PR when the code is ready. Thanks again.
